jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/server/Dispatcher.java

```diff
@@ -395,27 +455,13 @@ private static void applyAuthentication(String user, DataService dataService, En
      * Assumes, and does not check, that the action is a GSP action.
      */
     private static Operation gspOperation(HttpAction action, HttpServletRequest request) {
-        // Check enabled.
-        if ( isReadMethod(request) )
-            return GSP_R;
-        else
-            return GSP_RW;
+        return isReadMethod(request) ? GSP_R : GSP_RW;
     }
 
     /**
-     * Determine the {@link Operation} for a Quads operation. (GSP, except on the
-     * whole dataset).
-     * <p>
-     * Assumes, and does not check, that the action is a Quads action.
+     * Return whether request method is a "read" (GET or HEAD) or "write", modifying
+     * (POST, PUT, DELETE, PATCH)
```

Comment:
`ActionProcessor` has a switch on method type. TRACE isn't provided by any service IIRC. https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/TRACE questions whether it is used in browsers at all.

CONNECT doesn't get to the dispatcher.
(I'll add a definite rejection case to `ActionProcessor` to reject it rather than ignore it.)

It looks like OPTIONS works properly for everything except for multifunction endpoints for quads operations (this one case in `Dispatcher`).

We can't tell if it is "read" or "write" until the real operation is tried.

GSP_R vs GSP_RW is a weird case - one is a subset of the other.

So it gets "GSP_RW" which gives the full range of methods possible. Unfortunately, if only GSP_R is configured it will bounced because there is no implementation of GSP_RW.  Probably a corner case bug.

Having secure by-design read-only operation is more important than details of OPTIONS (which I never seen used on Fuseki - except by the test suite).

Reply:
PS

Overloaded endpoint, GSP_RW: OPTIONS works.

Overloaded endpoint, GSP_R: OPTIONS is rejected with "405 HTTP method not allowed: OPTIONS" which isn't illegal, just unhelpful.

The dispatcher in just for SPARQL - it's a full HTTP REST environment! So I might fix it just because.

Reply:
`OPTIONS` being rejected is most likely to be a problem for browser based UIs since it's used as a pre-flight request by browsers prior to sending the real request

Reply:
Thanks for the explanation & investigation @afs!

Reply:
@rvesse -- CORS pre-flight is handled before Fuseki.

https://github.com/apache/jena/blob/ea9406a6ca5be4e34c6e7db57e282947be529450/jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java#L1452

It is only non-CORS OPTIONS here.

Reply:
> @rvesse -- CORS pre-flight is handled before Fuseki.
> 
> https://github.com/apache/jena/blob/ea9406a6ca5be4e34c6e7db57e282947be529450/jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java#L1452
> 
> It is only non-CORS OPTIONS here.

I suspected as much, thanks for confirming